File: server/runtime/cues.py
"""
Cue language — the shared polish vocabulary every game emits to drive
audio, music, SFX, LED patterns, vibration motor, and any future
hardware feedback channel.

Why a cue layer rather than per-game `play("ding.mp3")` calls? Because
every game needs the same handful of beats (someone scored, round
about to end, match started, player eliminated). A cue is a semantic
name; the polish system decides which file plays, which LED palette
flashes, how long the motor buzzes. Asset swaps happen in one place
without touching game logic.

A game emits cues by attaching them to its StateUpdate.cues list. The
MatchManager folds the list into matches.snapshot, the TV diffs the
array across snapshot updates and fires onCue handlers per channel.
Firmware reads its own subset (LED + motor + buzzer) over the puck
HTTP path.

Adding a new cue:
  1. Add a value to Cue below with a one-line comment on intent.
  2. Add asset files at server/assets/<channel>/<cue_name>.<ext>.
  3. Update server/assets/<channel>/assets.yaml with attribution.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional


class CueChannel(str, Enum):
    """Which feedback layer the cue targets. Subscribers filter on
    channel so an audio-only TV ignores motor cues, etc."""

    VO = "vo"          # Host voice-over line (full sentence, ducks music)
    MUSIC = "music"    # Music bed transitions (start, stop, intensify)
    SFX = "sfx"        # Short ambient effect (ding, buzz, whoosh)
    LED = "led"        # WS2812B ring pattern on the puck
    MOTOR = "motor"    # Vibration motor pulse
    BUZZER = "buzzer"  # Piezo buzzer tone on the puck
    HAPTIC = "haptic"  # Catch-all for combined LED+motor sequences


class Cue(str, Enum):
    """Semantic cue names. The vocabulary is shared across all games so
    polish authors can map cue -> asset once and every game benefits.

    Naming convention: <subject>_<event>. SUBJECTs are PLAYER (the
    puck that triggered), MATCH (the whole match), ROUND, or
    TIMER. EVENTs are concrete moments.
    """

    # Match lifecycle
    MATCH_START = "match_start"
    MATCH_END = "match_end"
    MATCH_RESET = "match_reset"

    # Round structure (used by trivia / golf / racer)
    ROUND_START = "round_start"
    ROUND_END = "round_end"
    ROUND_TRANSITION = "round_transition"

    # Player events
    PLAYER_JOINED = "player_joined"
    PLAYER_LEFT = "player_left"
    PLAYER_CORRECT = "player_correct"      # Right answer, good shot, hit
    PLAYER_WRONG = "player_wrong"          # Wrong answer, miss
    PLAYER_TIMEOUT = "player_timeout"
    PLAYER_LOCKED_IN = "player_locked_in"  # Answer / shot committed
    PLAYER_ELIMINATED = "player_eliminated"
    PLAYER_VICTORY = "player_victory"

    # Score moments
    SCORE_GOLD = "score_gold"        # Top tier (Speed Pyramid <3s)
    SCORE_SILVER = "score_silver"
    SCORE_BRONZE = "score_bronze"
    SCORE_COMBO = "score_combo"      # Multiplier earned

    # Timer beats
    TIMER_TICK = "timer_tick"           # Each second remaining
    TIMER_WARNING = "timer_warning"     # 5 seconds left
    TIMER_EXPIRED = "timer_expired"


@dataclass
class CueEvent:
    """One cue instance fired during a state update.

    target: optional puck_index when the cue is player-specific. None
    means "broadcast" (whole table). Firmware can filter by target so
    only the relevant puck buzzes for player_correct.

    channels: which feedback layers should react. Default is "all" — the
    cue dispatcher decides what each layer's asset is. Authoring a
    cue with a specific channels tuple is the escape hatch for VO-only
    or motor-only beats.

    ts: server-side timestamp at emission. The TV uses this to dedupe
    cues across overlapping snapshot updates.

    payload: optional cue-specific data — score amount, player name,
    question category. Stays small (JSON-serialisable, <100B).
    """

    cue: Cue
    target: Optional[int] = None
    channels: Optional[tuple[CueChannel, ...]] = None
    ts: float = field(default_factory=time.time)
    payload: dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "cue": self.cue.value,
            "target": self.target,
            "channels": (
                [c.value for c in self.channels]
                if self.channels is not None
                else None
            ),
            "ts": self.ts,
            "payload": self.payload,
        }


# Convenience constructors so games can emit cues without verbose
# dataclass instantiation:
#
#     update = StateUpdate(state=..., cues=[
#         cue_player_correct(puck_index=3, payload={'tier': 'gold'}),
#         cue_score(puck_index=3, payload={'amount': 1000}),
#     ])


def cue_player_correct(puck_index: int, **payload) -> CueEvent:
    return CueEvent(cue=Cue.PLAYER_CORRECT, target=puck_index, payload=payload)


def cue_player_wrong(puck_index: int, **payload) -> CueEvent:
    return CueEvent(cue=Cue.PLAYER_WRONG, target=puck_index, payload=payload)


def cue_match_start(**payload) -> CueEvent:
    return CueEvent(cue=Cue.MATCH_START, payload=payload)


def cue_match_end(winner_index: Optional[int] = None, **payload) -> CueEvent:
    return CueEvent(
        cue=Cue.MATCH_END,
        target=winner_index,
        payload=payload,
    )


def cue_round_start(round_number: int, **payload) -> CueEvent:
    return CueEvent(
        cue=Cue.ROUND_START,
        payload={"round_number": round_number, **payload},
    )


def cue_timer_warning(seconds_left: int) -> CueEvent:
    return CueEvent(
        cue=Cue.TIMER_WARNING,
        payload={"seconds_left": seconds_left},
    )

File: TableWarsTV/src/lib/cues.ts
// Cue language mirror of server/runtime/cues.py. Keep these enum values
// in lockstep with the Python side; the runtime ships cue names in
// matches.snapshot.cues and the dispatcher matches on the string value.

export const Cue = {
  // Match lifecycle
  MATCH_START: 'match_start',
  MATCH_END: 'match_end',
  MATCH_RESET: 'match_reset',
  // Round structure
  ROUND_START: 'round_start',
  ROUND_END: 'round_end',
  ROUND_TRANSITION: 'round_transition',
  // Player events
  PLAYER_JOINED: 'player_joined',
  PLAYER_LEFT: 'player_left',
  PLAYER_CORRECT: 'player_correct',
  PLAYER_WRONG: 'player_wrong',
  PLAYER_TIMEOUT: 'player_timeout',
  PLAYER_LOCKED_IN: 'player_locked_in',
  PLAYER_ELIMINATED: 'player_eliminated',
  PLAYER_VICTORY: 'player_victory',
  // Score moments
  SCORE_GOLD: 'score_gold',
  SCORE_SILVER: 'score_silver',
  SCORE_BRONZE: 'score_bronze',
  SCORE_COMBO: 'score_combo',
  // Timer beats
  TIMER_TICK: 'timer_tick',
  TIMER_WARNING: 'timer_warning',
  TIMER_EXPIRED: 'timer_expired',
} as const;

export type CueName = (typeof Cue)[keyof typeof Cue];

export const CueChannel = {
  VO: 'vo',
  MUSIC: 'music',
  SFX: 'sfx',
  LED: 'led',
  MOTOR: 'motor',
  BUZZER: 'buzzer',
  HAPTIC: 'haptic',
} as const;

export type CueChannelName = (typeof CueChannel)[keyof typeof CueChannel];

export type CueEvent = {
  cue: CueName;
  target: number | null;
  channels: CueChannelName[] | null;
  // Per-match monotonic sequence stamped by the MatchManager. This is the
  // dedupe key — it only ever increases, unlike `ts`.
  seq: number;
  // Wall-clock emission time (seconds). Display/debug only; NOT used for
  // dedupe because NTP steps / container restarts can run it backwards.
  ts: number;
  payload: Record<string, unknown>;
};

// === Dispatcher ===
//
// The TV diffs snapshot.cues across updates so it only fires new cue
// emissions. Each cue carries a server-side `seq` — a per-match monotonic
// counter — and we keep the highest `seq` we've seen, treating anything
// at or below it as already dispatched. We deliberately do NOT key off
// the wall-clock `ts`: an NTP adjustment or container restart can make a
// later cue's ts smaller than an earlier one's, which would stick the
// watermark high and silently swallow every subsequent cue.
//
// Subscribers register by channel; the dispatcher fans cue events out
// to every subscriber whose channel is in the cue's channels list (or
// to all subscribers when channels is null, meaning "broadcast").
//
// Audio playback library is intentionally not wired yet — react-native
// audio choice (sound vs track-player) gets decided when the first
// real cue lands. For now, subscribers receive the typed event and the
// audio handler is a stub the polish system can swap in.

export type CueHandler = (event: CueEvent) => void;

// Sentinel below any real seq (which start at 0), so the first cue of a
// match is always fresh.
const NO_SEQ = -1;

export class CueDispatcher {
  private handlers = new Map<CueChannelName, Set<CueHandler>>();
  private lastSeq = NO_SEQ;

  on(channel: CueChannelName, handler: CueHandler): () => void {
    if (!this.handlers.has(channel)) {
      this.handlers.set(channel, new Set());
    }
    this.handlers.get(channel)!.add(handler);
    return () => this.handlers.get(channel)?.delete(handler);
  }

  // Call this every time matches.snapshot.cues changes. Cues at or below
  // the highest seq we've seen are skipped (already dispatched). A cue
  // missing `seq` falls back to its array index within this batch so a
  // pre-seq server snapshot still dispatches once rather than never.
  ingest(cues: CueEvent[]): void {
    if (!cues || cues.length === 0) return;
    // Resolve a dedupe seq per cue up front (server-stamped seq, or a
    // synthetic fallback keyed off the ORIGINAL index for a pre-seq
    // snapshot) so the post-filter index shift can't corrupt it.
    const withSeq = cues.map((cue, i) => ({
      cue,
      seq: typeof cue.seq === 'number' ? cue.seq : this.lastSeq + 1 + i,
    }));
    const fresh = withSeq.filter((x) => x.seq > this.lastSeq);
    if (fresh.length === 0) return;
    this.lastSeq = Math.max(...fresh.map((x) => x.seq));
    for (const { cue } of fresh) {
      const targetChannels =
        cue.channels && cue.channels.length > 0
          ? cue.channels
          : (Array.from(this.handlers.keys()) as CueChannelName[]);
      for (const ch of targetChannels) {
        const set = this.handlers.get(ch);
        if (!set) continue;
        for (const handler of set) handler(cue);
      }
    }
  }

  // Reset on new match so old cues don't carry over.
  reset(): void {
    this.lastSeq = NO_SEQ;
  }
}
